go/kbfs/simplefs/downloads.go

```diff
@@ -0,0 +1,353 @@
+	}
+	// could race but it should be rare enough so fine
+
+	err = os.Rename(srcPath, destPath)
+	if err != nil {
+		// Rename failed; assume it's a cross-partition move.
+		opid, err := m.k.SimpleFSMakeOpid(ctx)
+		if err != nil {
+			return "", err
+		}
+		err = m.k.SimpleFSMove(ctx, keybase1.SimpleFSMoveArg{
+			OpID: opid,
+			Src:  keybase1.NewPathWithLocal(srcPath),
+			Dest: keybase1.NewPathWithLocal(destPath),
+		})
+		if err != nil {
+			return "", err
+		}
+		err = m.k.SimpleFSWait(ctx, opid)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return destPath, nil
+}
+
+func (m *downloadManager) waitForDownload(ctx context.Context,
+	downloadID string, downloadPath string, done func(error)) {
+	d, err := m.getDownload(downloadID)
+	if err != nil {
+		done(err)
+		return
+	}
+	err = m.k.SimpleFSWait(ctx, d.opid)
+	if err != nil {
+		done(err)
+		return
+	}
+
+	var localPath string
+	if d.info.IsRegularDownload {
+		localPath, err = m.moveToDownloadFolder(ctx, downloadPath, d.info.Filename)
+		if err != nil {
+			done(err)
+			return
+		}
+	} else {
+		localPath = downloadPath
+	}
+	if err = m.updateDownload(downloadID, func(d download) download {
+		d.state.LocalPath = localPath
+		return d
+	}); err != nil {
+		done(err)
+		return
+	}
+	done(nil)
+}
+
+func (m *downloadManager) startDownload(
+	ctx context.Context, arg keybase1.SimpleFSStartDownloadArg) (
+	downloadID string, err error) {
+	opid, err := m.k.SimpleFSMakeOpid(ctx)
+	if err != nil {
+		return "", err
+	}
+	downloadID = strconv.FormatInt(time.Now().UnixNano(), 16)
+	downloadPath, filename, err := m.getDownloadPath(ctx, arg.Path, downloadID)
+	if err != nil {
+		return "", err
+	}
+	// TODO for dirs maybe we want zip instead?
+	err = m.k.SimpleFSCopyRecursive(ctx, keybase1.SimpleFSCopyRecursiveArg{
+		OpID: opid,
+		Src:  keybase1.NewPathWithKbfs(arg.Path),
+		Dest: keybase1.NewPathWithLocal(downloadPath),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	func() {
+		defer m.publisher.DownloadStatusChanged()
+		m.lock.Lock()
+		defer m.lock.Unlock()
+		m.downloads[downloadID] = download{
+			info: keybase1.DownloadInfo{
+				DownloadID:        downloadID,
+				Path:              arg.Path,
+				Filename:          filename,
+				StartTime:         keybase1.ToTime(time.Now()),
+				IsRegularDownload: arg.IsRegularDownload,
+			},
+			opid: opid,
+			state: keybase1.DownloadState{
+				DownloadID: downloadID,
+			},
+		}
+	}()
+
+	bgCtx, cancelBtCtx := m.makeContext()
+	done := func(err error) {
+		_ = m.updateDownload(downloadID, func(d download) download {
+			if d.state.Done || d.state.Canceled || len(d.state.Error) > 0 {
+				return d
+			}
+			if err == context.Canceled {
+				d.state.Canceled = true
+			} else if err != nil {
+				d.state.Error = err.Error()
+			} else {
+				d.state.EndEstimate = keybase1.ToTime(time.Now())
+				d.state.Progress = 1
+				d.state.Done = true
+			}
+			return d
+		})
+		cancelBtCtx()
+	}
+	go m.monitorDownload(bgCtx, opid, downloadID, done)
```

Comment:
A little nervous about these goroutines leaking, especially in a test scenario, but I guess we have other stuff like this in SimpleFS already so it's probably not much worse.

Reply:
Yeah we already have these in similar manner in simplefs. But perhaps we can clean them up if idle (idle as in no GetDownloadStatus calls) for more than e.g. 1min? Currently GUI should query the status every second (the frequency notification happens).

Reply:
Eh, I wouldn't change it for now, just something to keep in mind.